src/Formidable.js

```diff
@@ -178,40 +178,55 @@ class IncomingForm extends EventEmitter {
     return true;
   }
 
+  // returns a promise if no callback is provided
   async parse(req, cb) {
     this.req = req;
+    let promise;
 
     // Setup callback first, so we don't miss anything from data events emitted immediately.
-    if (cb) {
-      const callback = once(dezalgo(cb));
-      this.fields = {};
-      const files = {};
-
-      this.on('field', (name, value) => {
-        if (this.type === 'multipart' || this.type === 'urlencoded') {
-          if (!hasOwnProp(this.fields, name)) {
-            this.fields[name] = [value];
-          } else {
-            this.fields[name].push(value);
-          }
-        } else {
-          this.fields[name] = value;
-        }
+    if (!cb) {
+      let resolveRef;
+      let rejectRef;
+      promise = new Promise((resolve, reject) => {     
+        resolveRef = resolve;
+        rejectRef = reject;
       });
-      this.on('file', (name, file) => {
-        if (!hasOwnProp(files, name)) {
-          files[name] = [file];
+      cb = (err, fields, files) => {
+        if (err) {
+          rejectRef(err);
         } else {
-          files[name].push(file);
+          resolveRef([fields, files]);
         }
-      });
-      this.on('error', (err) => {
-        callback(err, this.fields, files);
-      });
-      this.on('end', () => {
-        callback(null, this.fields, files);
-      });
+      }
     }
+    const callback = once(dezalgo(cb));
+    this.fields = {};
+    const files = {};
+
+    this.on('field', (name, value) => {
+      if (this.type === 'multipart' || this.type === 'urlencoded') {
+        if (!hasOwnProp(this.fields, name)) {
```

Comment:
Just a thought, this if statement (and the next one) could be simplified to:

    const existingValues = this.fields[name] ?? [];
    this.fields[name] = [...existingValues, value]

Reply:
I know but  we still support  node  16 I think not sure if the new syntax  works there